File: test/nua.test.js
/* Copyright (c) 2018-2020 Richard Rodger and other contributors */
'use strict'

var Lab = require('@hapi/lab')
Lab = null != Lab.script ? Lab : require('hapi-lab-shim')

const Code = require('@hapi/code')

const lab = (exports.lab = Lab.script())
const describe = lab.describe
const it = lab.it
const expect = Code.expect

const Nua = require('..')

describe('nua', function () {
  it('happy', () => {
    var base = { a: { b: 1 } }
    var base_a = base.a
    var src = { a: { b: 2 } }

    Nua(base, src)
    expect(base).equal({ a: { b: 2 } })
    expect(base_a === base.a).true()
  })

  it('null-fields', () => {
    var base = { a: 1, b: null, c: 4, d: 5, f: null }
    var src = { a: 2, b: 3, c: null, e: 6, f: { g: 7 } }

    Nua(base, src)

    expect(base).equal({
      a: 2, // override
      b: 3, // override (null is just a value)
      c: null, // override (null is just a value)
      // d is removed as not present in src
      e: 6, // defined in src
      f: { g: 7 }, // override (null is just a value)
    })
  })

  it('depth', () => {
    var src = { a: 11, b: { c: 22 }, d: { e: { f: 33 } } }

    var base0 = { a: 1, b: { c: 2 }, d: { e: { f: 3 } } }
    var base0b = base0.b
    var base0d = base0.d
    var base0e = base0.d.e
    Nua(base0, src, { depth: 4 })
    expect(base0).equal(src)
    expect(base0b === base0.b).true()
    expect(base0d === base0.d).true()
    expect(base0e === base0.d.e).true()

    var base1 = { a: 1, b: { c: 2 }, d: { e: { f: 3 } } }
    Nua(base1, src, { depth: 3 })
    expect(base1).equal(src)

    var base2 = { a: 1, b: { c: 2 }, d: { e: { f: 3 } } }
    Nua(base2, src, { depth: 2 })
    expect(base2).equal({ a: 11, b: { c: 22 }, d: { e: { f: 3 } } })

    var base3 = { a: 1, b: { c: 2 }, d: { e: { f: 3 } } }
    var base3b = base3.b
    var base3d = base3.d
    var base3e = base3.d.e
    Nua(base3, src, { depth: 1 })
    expect(base3).equal({ a: 11, b: { c: 2 }, d: { e: { f: 3 } } })
    expect(base3b === base3.b).true()
    expect(base3d === base3.d).true()
    expect(base3e === base3.d.e).true()

    var base4 = { a: 1, b: { c: 2 }, d: { e: { f: 3 } } }
    Nua(base4, src, { depth: 0 })
    expect(base4).equal(src)
  })

  it('array', () => {
    var base = { a: [1, 2] }
    var base_a = base.a

    Nua(base, { a: [11, 22] })
    expect(base).equal({ a: [11, 22] })
    expect(base_a === base.a).true()

    Nua(base, { a: [11, 22, 33] })
    expect(base).equal({ a: [11, 22, 33] })
    expect(base_a === base.a).true()

    Nua(base, { a: [11] })
    expect(base).equal({ a: [11] })
    expect(base_a === base.a).true()

    Nua(base, { a: { b: 1 } })
    var a = base.a
    a.b = 1
    expect(base).equal({ a: a })
    expect(base_a === base.a).true()
  })

  it('object', () => {
    var base = { a: { b: 1, c: 2 } }
    var base_a = base.a

    Nua(base, { a: { b: 11, c: 22 } })
    expect(base).equal({ a: { b: 11, c: 22 } })
    expect(base_a === base.a).true()

    Nua(base, { a: { b: 11, c: 22, d: 33 } })
    expect(base).equal({ a: { b: 11, c: 22, d: 33 } })
    expect(base_a === base.a).true()

    Nua(base, { a: { b: 11 } })
    expect(base).equal({ a: { b: 11 } })
    expect(base_a === base.a).true()

    Nua(base, { a: { b: 11 }, c: { d: { e: 1 } } })
    expect(base).equal({ a: { b: 11 }, c: { d: { e: 1 } } })
  })

  it('deep', () => {
    var base = {
      a: { c: 1, d: { e: 3, f: [5, { h: 7 }] } },
      b: [2, [4], { g: [6] }],
    }
    var base_a = base.a
    var base_ad = base.a.d
    var base_adf = base.a.d.f
    var base_adf1 = base.a.d.f[1]
    var base_b = base.b
    var base_b1 = base.b[1]
    var base_b2 = base.b[2]
    var base_b2g = base.b[2].g

    Nua(null, null)
    Nua(null, 1)
    Nua(1, null)

    Nua(base, null)
    expect(base).equal({
      a: { c: 1, d: { e: 3, f: [5, { h: 7 }] } },
      b: [2, [4], { g: [6] }],
    })

    Nua(base, 1)
    expect(base).equal({
      a: { c: 1, d: { e: 3, f: [5, { h: 7 }] } },
      b: [2, [4], { g: [6] }],
    })

    Nua(
      base,
      { a: { c: {}, d: { e: 3, f: [5, { h: 7 }] } } },
      { preserve: true }
    )
    expect(base).equal({
      a: { c: {}, d: { e: 3, f: [5, { h: 7 }] } },
      b: [2, [4], { g: [6] }],
    })

    Nua(base, {
      a: { c: 11, d: { e: 33, f: [55, { h: 77 }] } },
      b: [22, [44], { g: [66] }],
    })
    expect(base).equal({
      a: { c: 11, d: { e: 33, f: [55, { h: 77 }] } },
      b: [22, [44], { g: [66] }],
    })
    expect(base_a === base.a).true()
    expect(base_ad === base.a.d).true()
    expect(base_adf === base.a.d.f).true()
    expect(base_adf1 === base.a.d.f[1]).true()
    expect(base_b === base.b).true()
    expect(base_b1 === base.b[1]).true()
    expect(base_b2 === base.b[2]).true()
    expect(base_b2g === base.b[2].g).true()

    Nua(base, { a: { c: 11, d: { e: 33, f: [55] } }, b: [22, [44], {}] })
    expect(base).equal({
      a: { c: 11, d: { e: 33, f: [55] } },
      b: [22, [44], {}],
    })
    expect(base_a === base.a).true()
    expect(base_ad === base.a.d).true()
    expect(base_adf === base.a.d.f).true()
    expect(base_b === base.b).true()
    expect(base_b1 === base.b[1]).true()
    expect(base_b2 === base.b[2]).true()

    Nua(base, { a: { c: 11, d: { e: 33 } }, b: [22, [44]] })
    expect(base).equal({ a: { c: 11, d: { e: 33 } }, b: [22, [44]] })
    expect(base_a === base.a).true()
    expect(base_ad === base.a.d).true()
    expect(base_b === base.b).true()
    expect(base_b1 === base.b[1]).true()

    Nua(base, { a: { c: 11, d: {} }, b: [22] })
    expect(base).equal({ a: { c: 11, d: {} }, b: [22] })
    expect(base_a === base.a).true()
    expect(base_ad === base.a.d).true()
    expect(base_b === base.b).true()

    Nua(base, { a: { c: 11 } })
    expect(base).equal({ a: { c: 11 } })
    expect(base_a === base.a).true()

    Nua(base, {})
    expect(base).equal({})

    var src = {
      a: { c: 111, d: { e: 333, f: [555, { h: 777 }] } },
      b: [222, [444], { g: [666] }],
    }
    var src_a = src.a
    var src_ad = src.a.d
    var src_adf = src.a.d.f
    var src_adf1 = src.a.d.f[1]
    var src_b = src.b
    var src_b1 = src.b[1]
    var src_b2 = src.b[2]
    var src_b2g = src.b[2].g

    Nua(base, src)
    expect(base).equal({
      a: { c: 111, d: { e: 333, f: [555, { h: 777 }] } },
      b: [222, [444], { g: [666] }],
    })
    expect(src_a === base.a).true()
    expect(src_ad === base.a.d).true()
    expect(src_adf === base.a.d.f).true()
    expect(src_adf1 === base.a.d.f[1]).true()
    expect(src_b === base.b).true()
    expect(src_b1 === base.b[1]).true()
    expect(src_b2 === base.b[2]).true()
    expect(src_b2g === base.b[2].g).true()

    Nua(base, { a: 1 })
    expect(base).equal({
      a: 1,
    })
  })

  it('setter', () => {
    var base = { a: 1, b: [2] }

    function s0(obj, key, val) {
      obj[key] = 'object' === typeof val ? val : val + 1
    }

    Nua(
      base,
      { a: 2, b: [3, 4], c: 5, d: { e: 6, g: [7], h: null }, f: null, h: [8] },
      { depth: 0, setter: s0 }
    )
    expect(base).equal({
      a: 3,
      b: [4, 5],
      c: 6,
      d: { e: 7, g: [8], h: null },
      f: null,
      h: [9],
    })
  })

  it('preserve', () => {
    var base = { a: 1, b: { c: 2 }, f: null }

    Nua(base, { d: 3 }, { preserve: true })
    expect(base).equal({ a: 1, b: { c: 2 }, d: 3, f: null })

    Nua(base, { b: { c: 4 } }, { preserve: true })
    expect(base).equal({ a: 1, b: { c: 4 }, d: 3, f: null })

    Nua(base, { f: { g: 5 } }, { preserve: true })
    expect(base).equal({ a: 1, b: { c: 4 }, d: 3, f: { g: 5 } })
  })
})

File: go/nua_test.go
package nua

import (
	"reflect"
	"testing"
)

func mapPtr(m map[string]any) uintptr  { return reflect.ValueOf(m).Pointer() }
func slicePtr(s []any) uintptr         { return reflect.ValueOf(s).Pointer() }

func assertEqual(t *testing.T, got, want any) {
	t.Helper()
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("not equal\n got: %#v\nwant: %#v", got, want)
	}
}

func TestHappy(t *testing.T) {
	base := map[string]any{"a": map[string]any{"b": 1}}
	baseA := base["a"].(map[string]any)
	src := map[string]any{"a": map[string]any{"b": 2}}

	Merge(base, src)
	assertEqual(t, base, map[string]any{"a": map[string]any{"b": 2}})
	if mapPtr(baseA) != mapPtr(base["a"].(map[string]any)) {
		t.Fatalf("inner map reference not preserved")
	}
}

func TestNullFields(t *testing.T) {
	base := map[string]any{"a": 1, "b": nil, "c": 4, "d": 5, "f": nil}
	src := map[string]any{"a": 2, "b": 3, "c": nil, "e": 6, "f": map[string]any{"g": 7}}

	Merge(base, src)

	assertEqual(t, base, map[string]any{
		"a": 2,
		"b": 3,
		"c": nil,
		"e": 6,
		"f": map[string]any{"g": 7},
	})
}

func TestDepth(t *testing.T) {
	src := map[string]any{
		"a": 11,
		"b": map[string]any{"c": 22},
		"d": map[string]any{"e": map[string]any{"f": 33}},
	}

	mkBase := func() map[string]any {
		return map[string]any{
			"a": 1,
			"b": map[string]any{"c": 2},
			"d": map[string]any{"e": map[string]any{"f": 3}},
		}
	}

	base0 := mkBase()
	base0b := base0["b"].(map[string]any)
	base0d := base0["d"].(map[string]any)
	base0e := base0d["e"].(map[string]any)
	Merge(base0, src, WithDepth(4))
	assertEqual(t, base0, src)
	if mapPtr(base0b) != mapPtr(base0["b"].(map[string]any)) {
		t.Fatalf("depth=4: b ref lost")
	}
	if mapPtr(base0d) != mapPtr(base0["d"].(map[string]any)) {
		t.Fatalf("depth=4: d ref lost")
	}
	if mapPtr(base0e) != mapPtr(base0["d"].(map[string]any)["e"].(map[string]any)) {
		t.Fatalf("depth=4: d.e ref lost")
	}

	base1 := mkBase()
	Merge(base1, src, WithDepth(3))
	assertEqual(t, base1, src)

	base2 := mkBase()
	Merge(base2, src, WithDepth(2))
	assertEqual(t, base2, map[string]any{
		"a": 11,
		"b": map[string]any{"c": 22},
		"d": map[string]any{"e": map[string]any{"f": 3}},
	})

	base3 := mkBase()
	base3b := base3["b"].(map[string]any)
	base3d := base3["d"].(map[string]any)
	base3e := base3d["e"].(map[string]any)
	Merge(base3, src, WithDepth(1))
	assertEqual(t, base3, map[string]any{
		"a": 11,
		"b": map[string]any{"c": 2},
		"d": map[string]any{"e": map[string]any{"f": 3}},
	})
	if mapPtr(base3b) != mapPtr(base3["b"].(map[string]any)) {
		t.Fatalf("depth=1: b ref lost")
	}
	if mapPtr(base3d) != mapPtr(base3["d"].(map[string]any)) {
		t.Fatalf("depth=1: d ref lost")
	}
	if mapPtr(base3e) != mapPtr(base3["d"].(map[string]any)["e"].(map[string]any)) {
		t.Fatalf("depth=1: d.e ref lost")
	}

	base4 := mkBase()
	Merge(base4, src, WithDepth(0))
	assertEqual(t, base4, src)
}

func TestArray(t *testing.T) {
	base := map[string]any{"a": []any{1, 2}}
	// preserve enough capacity so identity survives same-length and shrink merges
	base["a"] = append(make([]any, 0, 8), base["a"].([]any)...)
	baseA := base["a"].([]any)

	Merge(base, map[string]any{"a": []any{11, 22}})
	assertEqual(t, base, map[string]any{"a": []any{11, 22}})
	if slicePtr(baseA) != slicePtr(base["a"].([]any)) {
		t.Fatalf("same-length merge: slice array ref lost")
	}

	Merge(base, map[string]any{"a": []any{11, 22, 33}})
	assertEqual(t, base, map[string]any{"a": []any{11, 22, 33}})
	if slicePtr(baseA) != slicePtr(base["a"].([]any)) {
		t.Fatalf("grow within cap: slice array ref lost")
	}

	Merge(base, map[string]any{"a": []any{11}})
	assertEqual(t, base, map[string]any{"a": []any{11}})
	if slicePtr(baseA) != slicePtr(base["a"].([]any)) {
		t.Fatalf("shrink: slice array ref lost")
	}

	// type change: src.a is a map, base.a is a slice — Go can't carry index
	// keys on a slice the way JS can on an array, so the value is replaced.
	Merge(base, map[string]any{"a": map[string]any{"b": 1}})
	assertEqual(t, base, map[string]any{"a": map[string]any{"b": 1}})
}

func TestObject(t *testing.T) {
	base := map[string]any{"a": map[string]any{"b": 1, "c": 2}}
	baseA := base["a"].(map[string]any)

	Merge(base, map[string]any{"a": map[string]any{"b": 11, "c": 22}})
	assertEqual(t, base, map[string]any{"a": map[string]any{"b": 11, "c": 22}})
	if mapPtr(baseA) != mapPtr(base["a"].(map[string]any)) {
		t.Fatalf("a ref lost after equal-shape merge")
	}

	Merge(base, map[string]any{"a": map[string]any{"b": 11, "c": 22, "d": 33}})
	assertEqual(t, base, map[string]any{"a": map[string]any{"b": 11, "c": 22, "d": 33}})
	if mapPtr(baseA) != mapPtr(base["a"].(map[string]any)) {
		t.Fatalf("a ref lost after key-add merge")
	}

	Merge(base, map[string]any{"a": map[string]any{"b": 11}})
	assertEqual(t, base, map[string]any{"a": map[string]any{"b": 11}})
	if mapPtr(baseA) != mapPtr(base["a"].(map[string]any)) {
		t.Fatalf("a ref lost after key-remove merge")
	}

	Merge(base, map[string]any{
		"a": map[string]any{"b": 11},
		"c": map[string]any{"d": map[string]any{"e": 1}},
	})
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"b": 11},
		"c": map[string]any{"d": map[string]any{"e": 1}},
	})
}

func TestDeep(t *testing.T) {
	base := map[string]any{
		"a": map[string]any{"c": 1, "d": map[string]any{"e": 3, "f": []any{5, map[string]any{"h": 7}}}},
		"b": []any{2, []any{4}, map[string]any{"g": []any{6}}},
	}
	// expand caps so slice identity survives length changes
	withCap := func(s []any, c int) []any { return append(make([]any, 0, c), s...) }
	a := base["a"].(map[string]any)
	d := a["d"].(map[string]any)
	d["f"] = withCap(d["f"].([]any), 8)
	base["b"] = withCap(base["b"].([]any), 8)
	bSlice := base["b"].([]any)
	bSlice[2].(map[string]any)["g"] = withCap(bSlice[2].(map[string]any)["g"].([]any), 8)

	baseA := base["a"].(map[string]any)
	baseAD := baseA["d"].(map[string]any)
	baseADF := baseAD["f"].([]any)
	baseADF1 := baseADF[1].(map[string]any)
	baseB := base["b"].([]any)
	baseB1 := baseB[1].([]any)
	baseB2 := baseB[2].(map[string]any)
	baseB2G := baseB2["g"].([]any)

	// nil and scalar src — base unchanged
	Merge(nil, nil)
	Merge(nil, 1)
	Merge(1, nil)

	Merge(base, nil)
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": 1, "d": map[string]any{"e": 3, "f": []any{5, map[string]any{"h": 7}}}},
		"b": []any{2, []any{4}, map[string]any{"g": []any{6}}},
	})

	Merge(base, 1)
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": 1, "d": map[string]any{"e": 3, "f": []any{5, map[string]any{"h": 7}}}},
		"b": []any{2, []any{4}, map[string]any{"g": []any{6}}},
	})

	Merge(base, map[string]any{"a": map[string]any{"c": map[string]any{}, "d": map[string]any{"e": 3, "f": []any{5, map[string]any{"h": 7}}}}}, WithPreserve(true))
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": map[string]any{}, "d": map[string]any{"e": 3, "f": []any{5, map[string]any{"h": 7}}}},
		"b": []any{2, []any{4}, map[string]any{"g": []any{6}}},
	})

	Merge(base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{"e": 33, "f": []any{55, map[string]any{"h": 77}}}},
		"b": []any{22, []any{44}, map[string]any{"g": []any{66}}},
	})
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{"e": 33, "f": []any{55, map[string]any{"h": 77}}}},
		"b": []any{22, []any{44}, map[string]any{"g": []any{66}}},
	})
	if mapPtr(baseA) != mapPtr(base["a"].(map[string]any)) {
		t.Fatalf("a ref lost")
	}
	if mapPtr(baseAD) != mapPtr(base["a"].(map[string]any)["d"].(map[string]any)) {
		t.Fatalf("a.d ref lost")
	}
	if slicePtr(baseADF) != slicePtr(base["a"].(map[string]any)["d"].(map[string]any)["f"].([]any)) {
		t.Fatalf("a.d.f ref lost")
	}
	if mapPtr(baseADF1) != mapPtr(base["a"].(map[string]any)["d"].(map[string]any)["f"].([]any)[1].(map[string]any)) {
		t.Fatalf("a.d.f[1] ref lost")
	}
	if slicePtr(baseB) != slicePtr(base["b"].([]any)) {
		t.Fatalf("b ref lost")
	}
	if slicePtr(baseB1) != slicePtr(base["b"].([]any)[1].([]any)) {
		t.Fatalf("b[1] ref lost")
	}
	if mapPtr(baseB2) != mapPtr(base["b"].([]any)[2].(map[string]any)) {
		t.Fatalf("b[2] ref lost")
	}
	if slicePtr(baseB2G) != slicePtr(base["b"].([]any)[2].(map[string]any)["g"].([]any)) {
		t.Fatalf("b[2].g ref lost")
	}

	Merge(base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{"e": 33, "f": []any{55}}},
		"b": []any{22, []any{44}, map[string]any{}},
	})
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{"e": 33, "f": []any{55}}},
		"b": []any{22, []any{44}, map[string]any{}},
	})

	Merge(base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{"e": 33}},
		"b": []any{22, []any{44}},
	})
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{"e": 33}},
		"b": []any{22, []any{44}},
	})

	Merge(base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{}},
		"b": []any{22},
	})
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": 11, "d": map[string]any{}},
		"b": []any{22},
	})

	Merge(base, map[string]any{"a": map[string]any{"c": 11}})
	assertEqual(t, base, map[string]any{"a": map[string]any{"c": 11}})

	Merge(base, map[string]any{})
	assertEqual(t, base, map[string]any{})

	src := map[string]any{
		"a": map[string]any{"c": 111, "d": map[string]any{"e": 333, "f": []any{555, map[string]any{"h": 777}}}},
		"b": []any{222, []any{444}, map[string]any{"g": []any{666}}},
	}
	srcA := src["a"].(map[string]any)
	srcAD := srcA["d"].(map[string]any)
	srcADF := srcAD["f"].([]any)
	srcADF1 := srcADF[1].(map[string]any)
	srcB := src["b"].([]any)
	srcB1 := srcB[1].([]any)
	srcB2 := srcB[2].(map[string]any)
	srcB2G := srcB2["g"].([]any)

	// when base lacks a key, merge installs the src reference directly
	Merge(base, src)
	assertEqual(t, base, map[string]any{
		"a": map[string]any{"c": 111, "d": map[string]any{"e": 333, "f": []any{555, map[string]any{"h": 777}}}},
		"b": []any{222, []any{444}, map[string]any{"g": []any{666}}},
	})
	if mapPtr(srcA) != mapPtr(base["a"].(map[string]any)) {
		t.Fatalf("base.a should now point at src.a")
	}
	if mapPtr(srcAD) != mapPtr(base["a"].(map[string]any)["d"].(map[string]any)) {
		t.Fatalf("base.a.d should now point at src.a.d")
	}
	if slicePtr(srcADF) != slicePtr(base["a"].(map[string]any)["d"].(map[string]any)["f"].([]any)) {
		t.Fatalf("base.a.d.f should now point at src.a.d.f")
	}
	if mapPtr(srcADF1) != mapPtr(base["a"].(map[string]any)["d"].(map[string]any)["f"].([]any)[1].(map[string]any)) {
		t.Fatalf("base.a.d.f[1] should now point at src.a.d.f[1]")
	}
	if slicePtr(srcB) != slicePtr(base["b"].([]any)) {
		t.Fatalf("base.b should now point at src.b")
	}
	if slicePtr(srcB1) != slicePtr(base["b"].([]any)[1].([]any)) {
		t.Fatalf("base.b[1] should now point at src.b[1]")
	}
	if mapPtr(srcB2) != mapPtr(base["b"].([]any)[2].(map[string]any)) {
		t.Fatalf("base.b[2] should now point at src.b[2]")
	}
	if slicePtr(srcB2G) != slicePtr(base["b"].([]any)[2].(map[string]any)["g"].([]any)) {
		t.Fatalf("base.b[2].g should now point at src.b[2].g")
	}

	Merge(base, map[string]any{"a": 1})
	assertEqual(t, base, map[string]any{"a": 1})
}

func TestSetter(t *testing.T) {
	base := map[string]any{"a": 1, "b": []any{2}}

	// In JS, typeof null === 'object', so the JS setter writes null through
	// unchanged. Match that here: nil and containers pass through, ints get
	// incremented.
	setter := func(parent any, key any, val any) {
		set := func(v any) {
			switch p := parent.(type) {
			case map[string]any:
				p[key.(string)] = v
			case []any:
				p[key.(int)] = v
			}
		}
		switch v := val.(type) {
		case nil:
			set(nil)
		case map[string]any, []any:
			set(v)
		case int:
			set(v + 1)
		default:
			set(v)
		}
	}

	src := map[string]any{
		"a": 2,
		"b": []any{3, 4},
		"c": 5,
		"d": map[string]any{"e": 6, "g": []any{7}, "h": nil},
		"f": nil,
		"h": []any{8},
	}
	Merge(base, src, WithDepth(0), WithSetter(setter))
	assertEqual(t, base, map[string]any{
		"a": 3,
		"b": []any{4, 5},
		"c": 6,
		"d": map[string]any{"e": 7, "g": []any{8}, "h": nil},
		"f": nil,
		"h": []any{9},
	})
}

func TestPreserve(t *testing.T) {
	base := map[string]any{"a": 1, "b": map[string]any{"c": 2}, "f": nil}

	Merge(base, map[string]any{"d": 3}, WithPreserve(true))
	assertEqual(t, base, map[string]any{
		"a": 1,
		"b": map[string]any{"c": 2},
		"d": 3,
		"f": nil,
	})

	Merge(base, map[string]any{"b": map[string]any{"c": 4}}, WithPreserve(true))
	assertEqual(t, base, map[string]any{
		"a": 1,
		"b": map[string]any{"c": 4},
		"d": 3,
		"f": nil,
	})

	Merge(base, map[string]any{"f": map[string]any{"g": 5}}, WithPreserve(true))
	assertEqual(t, base, map[string]any{
		"a": 1,
		"b": map[string]any{"c": 4},
		"d": 3,
		"f": map[string]any{"g": 5},
	})
}
